### ISSUE TITLE
Embeddings and centroids of classification models

### ISSUE THREAD

**Issue by pheersaquoff** [ultralytics/ultralytics#3269]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Is there a way to get class centroids in embedding space for classification models? And an embedding of a query image

**Comment by glenn-jocher** [comment]
Sure, you can get class centroids in the embedding space as well as embeddings for query images. However, this requires that your model is designed to output embeddings. This can be either a standalone model for generating embeddings (e.g., a siamese network, a triplet network, or a model trained with contrastive loss) or a standard classification model where you take the output of the penultimate layer as an "embedding" of the input.

Here's a general method to get class centroids and query image embeddings:

**Class Centroids:**

Class centroids are the average of the embeddings of all instances in a class. To calculate them:

1. First, pass all your training data (or a representative sample) through the model to get the embeddings.

2. Group the embeddings by class.

3. For each class, calculate the average (centroid) of the embeddings for that class. This is the class centroid.

**Query Image Embeddings:**

To get an embedding for a query image, just pass the query image through the model.

Here's some pseudocode:

```python
def get_embedding(model, dataloader):
    model.eval()
    embeddings = []
    labels = []
    with torch.no_grad():
        for inputs, targets in dataloader:
            inputs = inputs.to(device)
            outputs = model(inputs)
            embeddings.append(outputs)
            labels.append(targets)
    return torch.cat(embeddings), torch.cat(labels)

def get_centroids(embeddings, labels):
    centroids = {}
    for unique_label in labels.unique():
        mask = (labels == unique_label)
        centroids[unique_label.item()] = embeddings[mask].mean(dim=0)
    return centroids

# Get embeddings for all data
embeddings, labels = get_embedding(model, dataloader)

# Calculate centroids
centroids = get_centroids(embeddings, labels)

# Get embedding for a query image
query_image = ... # Load your query image
query_embedding = model(query_image)
```

This assumes that `model` is your model and `dataloader` is a PyTorch DataLoader that loads your data. Also, make sure that your model is in evaluation mode (`model.eval()`) and that you're handling device placement correctly (i.e., moving data to the same device as your model).

Remember, the specifics of how to implement this can vary depending on the structure of your model and how exactly you've implemented your training process.
